### ISSUE TITLE
🙏 Puzzled with ExpressiveCodeEngine

### ISSUE THREAD

**Issue by PassionZale** [expressive-code/expressive-code#219]
I'm using `rehype-expressive-code` as mdx plugins to highlight Code Blocks in `*.mdx`, it works better, like this:

```ts
// vite.config.ts
import type { UserConfig } from "vite";
import react from "@vitejs/plugin-react-swc";
import mdx from "@mdx-js/rollup";
import rehypeExpressiveCode, {
  RehypeExpressiveCodeOptions,
} from "rehype-expressive-code";

const rehypeExpressiveCodeOptions: RehypeExpressiveCodeOptions = {
  frames: {
    showCopyToClipboardButton: false,
  },
};

export default (): UserConfig => {
  plugins: [
    {
      ...mdx({
        remarkPlugins: [],
        rehypePlugins: [[rehypeExpressiveCode, rehypeExpressiveCodeOptions]],
      }),
    },
    react(),
  ]
}
```

you provide `astro-expressive-code`, so is easy to hightlight `filePath?raw` in `Astro`, like this:

```mdx
import { Code } from 'astro-expressive-code/components';
import importedCode from '/src/env.d.ts?raw';

<Code code={importedCode} lang="ts" title="src/env.d.ts" />
```

but I'm not use `expressive-code` in `Astro`, I just use it in `Vite + React`, so I want to implement a `React` component like `astro-expressive-code/components/Code`.

I find an example in [@expressive-code/core
](https://github.com/expressive-code/expressive-code/blob/main/packages/%40expressive-code/core/README.md).

and try it in my component, such like:

```tsx
import { useEffect, useState } from "react";
import { ExpressiveCodeEngine } from "@expressive-code/core";
import { toHtml } from "@expressive-code/core/hast";

const Code = () => {
  const [code, setCode] = useState("");

  useEffect(() => {
    renderRaw();
  }, []);

  const renderRaw = async () => {
    const ec = new ExpressiveCodeEngine({
      plugins: [
      ],
    });

    const baseStyles = await ec.getBaseStyles();
    const themeStyles = await ec.getThemeStyles();

    const renderResult = await ec.render({
      code: 'console.log("Hello world!")',
      language: "tsx",
    });

    console.dir({
      baseStyles,
      themeStyles,
      blockStyles: renderResult.styles,
      blockHtml: toHtml(renderResult.renderedGroupAst),
    });

    setCode(toHtml(renderResult.renderedGroupAst));
  };

  return <div dangerouslySetInnerHTML={{ __html: code }} />;
};

export default Code;
```

and use it in my `.mdx`：

![image](https://github.com/expressive-code/expressive-code/assets/12482130/8f2cf60d-97c3-4fbd-93a8-9dbdbb720980)

the `<Code />` component render success, but highlight missing:

![image](https://github.com/expressive-code/expressive-code/assets/12482130/d7bd7e99-5039-45b8-9bed-2fcd9d4dbaa9)

Code Blocks in `*.mdx` highlight success, and auto add `style` & `script` & `css vars`, render diffrent DOM structure：

![image](https://github.com/expressive-code/expressive-code/assets/12482130/a9e26a90-4909-4850-8101-609b338f261e)

**read your documents, It means I need to add `baseStyles` `themeStyles` `jsModules` to DOM?**

I have used `Prism` or `Shiki`, these packages will provide `highlight function`, it's easy to use like:

```ts
Prism.highlight(props.code, prismGrammar, props.language)
```

I read `expressive-code` source code and all the documents, no `highlight function` provide, the only `function` I found is `toHtml()`.

and there's no example to show how to add `baseStyles` `themeStyles` `jsModules` to DOM,

**🙏 Can you give some React example like `astro-expressive-code/components/Code`**,

I'm writing a React Component Library Documents, so I need to render `*.tsx?raw` and highlight `fileRawData`

**Comment by PassionZale** [comment]
I have also tried to add `baseStyles` `themeStyles` `jsModules`, highlight doesn't work too:

```diff
import { useEffect, useState } from "react";
import { ExpressiveCodeEngine } from "@expressive-code/core";
import { toHtml } from "@expressive-code/core/hast";

const Code = () => {
  const [code, setCode] = useState("");

  useEffect(() => {
    renderRaw();
  }, []);

  const renderRaw = async () => {
    const ec = new ExpressiveCodeEngine({
      plugins: [
      ],
    });

    const baseStyles = await ec.getBaseStyles();
    const themeStyles = await ec.getThemeStyles();

    const renderResult = await ec.render({
      code: 'console.log("Hello world!")',
      language: "tsx",
    });

    console.dir({
      baseStyles,
      themeStyles,
      blockStyles: renderResult.styles,
      blockHtml: toHtml(renderResult.renderedGroupAst),
    });

-    setCode(toHtml(renderResult.renderedGroupAst));
    
+    setCode(
+      `<style>${themeStyles}${baseStyles}</style>
+	<script type="module">${jsModules}</script>
+	${toHtml(renderResult.renderedGroupAst)}
+      `
+    )
+  };

  return <div dangerouslySetInnerHTML={{ __html: code }} />;
};

export default Code;
```


![image](https://github.com/expressive-code/expressive-code/assets/12482130/c50c5254-ac54-434a-857f-c1d023c41921)

**Comment by hippotastic** [comment]
Hello! Wow, you did a lot of experimentation there already! Thank you for reading the docs and testing various approaches already.

The likely reason your styles don't work is that it's too late when React is hydrating to insert links to CSS or JS modules into the HTML source code. The browser will not load them anymore at this point, as these links need to be a part of the server-side output to be picked up.

I'm not using React like this myself, so I can't really provide a working solution unfortunately. I'd also like to mention that Expressive Code is not designed to be used on the client, but only on the server. Using it from React frontend code will result in a huge client-side bundle.

The Code component I provide for Astro is different - even though it might look similar to the React version you're trying to achieve, it is fully server-rendered and therefore does not have any of the issues like bundle size or CSS/JS module linking problems.

In summary, I think that Expressive Code is probably the wrong tool for this kind of job right now, and if you want client-side highlighting with React, you'd be better off using any of the excellent alternatives that are designed for this use case. You already mentioned Prism and Shiki, and I'd suggest using these.

Sorry not to be able to offer more help here!

**Comment by PassionZale** [comment]
Thks your great explains!

> I'd also like to mention that Expressive Code is not designed to be used on the client, but only on the server. 

so I known why code highlihgt work in `*.mdx` but not work in `Component`:

- `*.mdx` is parsed by `@mdx-js/rollup`, server-side render;
- `Component` is client-side render;

> In summary, I think that Expressive Code is probably the wrong tool for this kind of job right now, and if you want client-side highlighting with React, you'd be better off using any of the excellent alternatives that are designed for this use case. You already mentioned Prism and Shiki, and I'd suggest using these.

I had used `react-syntax-highlighter` to review `<Code />` and it works now:

![image](https://github.com/expressive-code/expressive-code/assets/12482130/e6c36ca2-5639-4a22-926a-584f3942cfa5)

`expressive-code` is a great repo 👍

I use it and `Starlight` create the Company Documentation Center, my colleagues think it's great.
